lib/SILOptimizer/Utils/InstOptUtils.cpp

```diff
@@ -994,38 +1005,71 @@ swift::castValueToABICompatibleType(SILBuilder *builder, SILLocation loc,
 
     // Unwrap the original optional value.
     auto *someDecl = builder->getASTContext().getOptionalSomeDecl();
-    auto *noneBB = builder->getFunction().createBasicBlock();
-    auto *someBB = builder->getFunction().createBasicBlock();
     auto *curBB = builder->getInsertionPoint()->getParent();
-
     auto *contBB = curBB->split(builder->getInsertionPoint());
-    contBB->createPhiArgument(destTy, OwnershipKind::Owned);
+    auto *someBB = builder->getFunction().createBasicBlockAfter(curBB);
+    auto *noneBB = builder->getFunction().createBasicBlockAfter(someBB);
+
+    auto *phi = contBB->createPhiArgument(destTy, value.getOwnershipKind());
+    if (phi->getOwnershipKind() == OwnershipKind::Guaranteed) {
+      auto createEndBorrow = [&](SILBasicBlock::iterator insertPt) {
+        builder->setInsertionPoint(insertPt);
+        builder->createEndBorrow(loc, phi);
+      };
+      for (SILInstruction *user : usePoints) {
```

Comment:
If you want there is an insert after utility that does this. I am fine with that in a follow on commit though.

Reply:
That utility is fine but the builder is being passed in here and we can't be sure that the caller isn't relying on everything going through that builder.